### ISSUE TITLE
[MNG-5085] Add a CLI option to ignore missing modules

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7164]
**[Stephan Pauxberger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=paux)** opened **[MNG-5085](https://issues.apache.org/jira/browse/MNG-5085?redirect=false)** and commented

Using SVN for a rather big project, we tend to use SVN sparse checkouts, i.e. we do not checkout the whole project. Example:

Full Project (as in Repository):
Parent
pom.xml (contains A and B as modules)
--> A
pom.xml
--> B
pom.xml

Now, do a checkout (svn co xxx --depth children; svn update --set-depth inifity A)

Working Copy:
Parent
pom.xml (contains A and B as modules)
--> A
pom.xml
--> B (no pom!!, since we only did a sparse checkout)

Now, this setup is not buildable, since maven complains (rightfully) about a missing pom for B.

What I propose is an option to change this behaviour with a command-line option (-imm, --ignore-missing-modules) that would simply ignore missing modules during pom resolution.

---

**Issue Links:**
- [MNG-5230](https://issues.apache.org/jira/browse/MNG-5230) Command line option to exclude modules from reactor

1 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

In this scenario, you should use profiles to express what modules can be safely "turned off"

**Comment by jira-importer** [comment]
**[Stephan Pauxberger](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=paux)** commented

Our project consists of about 30 modules. To be able to individually combine "active" and "not active" modules, we would need a profile for each of the modules, which would be quite extreme.

Alternately, you could specify the projects to build using "-pl", however even the projects that you do not want to be built need to be present, making the use of sparse checkouts impossible.

My proposal is a way similar to parent projects (i.e. if the parent project is present, that one is used, else the parent is retrieved from repo).

This would allow the following workflow:

- I have a task to work upon, concerning modules A12, A13 and A23.
- I do a sparse checkout of the master project
- I widen the checkout to include A12, A13 and A23 (these steps take about 1/10th of the time it would take to do a complete checkout!)
- I work upon my task, whenever I need a build I just build the whole scaffold project using "mvn -imm install", which builds only parent an my three modules
- I realize my task also concerns A7, so I widen my working copy to include that as well
- nothing changes in respect to compilation etc.., what is there gets built, what is missing is retrieved from repo.

Major benefit of this workflow:

Since my working copy is way smaller EVERY subversion operation is faster by magnitudes.

**Comment by jira-importer** [comment]
**[Ivan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebann)** commented

We do have a multi module projects with more than 40 submodules too.

We can use profiles to make modules optional but that will lead to a long and complex pom.xml.

That would be nice to be able to say "don't fail if modules are not existing".

**Comment by jira-importer** [comment]
**[Ivan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ebann)** commented

Well in my case this is more in order to provide "sparse" workspaces to developers.

I mean, we have some core modules and many plugin modules.
Some developers are not interested in having the core modules (and in some cases we don't want them to have it, for security reasons).
But they still need to build the whole application regardless they don't have the sources for the core project or for non related plugins.

During the build, missing modules needed (dependency) will be downloaded from the repository.
And missing modules that are not present and not needed won't be packaged.
(An even better solution might be to tell maven that modules which can't be built should be searched in the local/remote repository)

In other words, to have the same behavior for dependency resolutions than for modules resolutions.

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

Ivan, I don't think this option is really necessary for Maven. Based on your description, your issue seems purely process related. Your developers should be able to check out parts of the tree as necessary but you want to have the rest of tree already deployed to your corporate repository. Regarding building the whole thing, that's really a task for a tool like Hudson.

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

Ivan, have you considered private branches for your developers? It is a great help on large projects because large projects cannot tolerate work that isn't ready for integration. This will then allow Hudson to build "the whole thing" and test/demo.

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

I stumbled upon an article today that tangentially regards this issue. It is mostly FUD but it also echos some of the recent sentiments here regarding the difficulty of modules/nesting. Just linking to it to help the discussion: http://buzdin.blogspot.com/2012/08/maven-modules-considered-harmful.html
